unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp

```diff
@@ -0,0 +1,918 @@
+
+    THEN("The return type should be correct")
+    {
+      require_type::require_pointer(
+        function_call.return_type(),
+        symbol_typet(class_prefix + "$GenericInnerClass"));
+      require_type::require_java_generic_type(
+        function_call.return_type(),
+        {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
+         {require_type::type_parameter_kindt::Inst,
+          "java::Interface_Implementation"}});
+    }
+  }
+  THEN(
+    "Ret method 3 should return a "
+    "GenericClassWithGenericInnerClasses$GenericInnerClass")
+  {
+    const std::string func_name = ".ret_method3";
+    const std::string func_descriptor =
+      ":()LGenericClassWithGenericInnerClasses$GenericInnerClass;";
+    const std::string process_func_name =
+      class_prefix + func_name + func_descriptor;
+
+    REQUIRE(new_symbol_table.has_symbol(process_func_name));
+    const symbolt &function_symbol =
+      new_symbol_table.lookup_ref(process_func_name);
+
+    const code_typet &function_call =
+      require_type::require_code(function_symbol.type, 1);
+
+    THEN("The return type should be correct")
+    {
+      require_type::require_pointer(
+        function_call.return_type(),
+        symbol_typet(class_prefix + "$GenericInnerClass"));
+      require_type::require_java_generic_type(
+        function_call.return_type(),
+        {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
+         {require_type::type_parameter_kindt::Var, class_prefix + "::T"}});
+    }
+  }
+  THEN(
+    "Ret method 4 should return a GenericClassWithGenericInnerClasses"
+    "$GenericInnerClass$DoublyNestedInnerClass")
+  {
+    const std::string func_name = ".ret_method4";
+    const std::string func_descriptor =
+      ":()LGenericClassWithGenericInner"
+      "Classes$GenericInnerClass$DoublyNestedInnerClass;";
+    const std::string process_func_name =
+      class_prefix + func_name + func_descriptor;
+
+    REQUIRE(new_symbol_table.has_symbol(process_func_name));
+    const symbolt &function_symbol =
+      new_symbol_table.lookup_ref(process_func_name);
+
+    const code_typet &function_call =
+      require_type::require_code(function_symbol.type, 1);
+
+    THEN("The return type should be correct")
+    {
+      require_type::require_pointer(
+        function_call.return_type(),
+        symbol_typet(
+          class_prefix + "$GenericInnerClass$DoublyNestedInnerClass"));
+      require_type::require_java_generic_type(
+        function_call.return_type(),
+        {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
+         {require_type::type_parameter_kindt::Inst,
+          "java::Interface_Implementation"}});
+    }
+  }
+  THEN(
+    "Ret method 5 should return a GenericClassWithGenericInnerClasses"
+    "$GenericInnerClass$DoublyNestedInnerClass")
+  {
+    const std::string func_name = ".ret_method5";
+    const std::string func_descriptor =
+      ":()LGenericClassWithGenericInner"
```

Comment:
Minor nit: again, probably break this string after the `Classes$` part. I'm guessing CLion broke these up in an odd way without making it too obvious...